### PR TITLE
feat: mobile support

### LINKED ISSUE #33: Mobile Android/ iOS Support
**Tell us about the task you want to perform and are unable to do so because the feature is not available**
I would like to add support for Mobile Platforms i.e. Android, iOS this would allow to test APIs on-the-go. It might seem excessive and unnecessary but this kind of agility that enables everyone to test stuff on-the-go, having the same familiar UX and set of features providing continuity, with the added benefit of having a shared open-source codebase powered by flutter is what could potentially set this app apart from other competition.

**Describe the solution/feature you'd like us to add**
My proposal is building a preliminary mobile version of the app with a simplified UI/UX that can empower users to test APIs and endpoints in the simplest way possible on-the-go. This could potentially be utilized in adding future features like integrating a cross-platform sync; providing continuity for signed-up users.

**Approach**
This is obviously a fairly big feature to be added, for that my approach would be to start
1. Designing an interface with the bare minimum of functional features coupled with the same beautiful design, (no changes, just adapting for smaller screens)
2. Iteratively develop/ adapt code for the new platforms (starting with Android because it's usually more accessible)
3. Test and add more features while maintaining this simple design language.

### PR DESCRIPTION
**Pull request by DenserMeerkat**
## PR Description

 Implementing the [Figma Link](https://www.figma.com/file/6gsFPwN1NRv7XuVpXZ9FrP/apidash_mobile-DenserMeerkat?type=design&node-id=0%3A1&mode=design&t=uumUAO9f7BfcWAag-1) design with maximum (almost all) available components, to discuss the design decisions.

https://github.com/foss42/apidash/assets/95911940/46c27c75-d4fa-4593-bf0c-cb84541170b3

## Related Issues

- Related Issue #33 
- Closes #33 

### Checklist
- [X] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [X] I have run the tests (`flutter test`) and all tests are passing

## Added/updated tests?

- [ ] Yes
- [X] No, and this is why: the features are still in implementation phase.
- [ ] I need help with writing tests

### CODE REVIEW

**Comment by DenserMeerkat** [comment]
Currently, I have tried to make the most of existing widgets by adding `kIsMobile` conditions. Are we supposed to go full on that much or can we create some second level layout widgets(widgets that composite low level widgets) specific to mobile.
Expample: `RequestEditor()`

**Comment by animator** [comment]
@DenserMeerkat Thanks for the PR. We will review it and get back.

**Comment by animator** [comment]
You are going in the right direction. I recommend you to continue thinking about the end user and improve the UI/UX to make his life easier and the app more intuitive to use.

Some feedback points:
- Click on rename button. Enter nothing and press OK. You get an error screen.
227 rename
227 error

- Type something in the body text field. click outside the textfield. You will not be able to unfocus and hide the keyboard to access the send button.  

227 textbox

- Add a lot of header rows now scroll. A strange error happens. Investigate.

https://github.com/foss42/apidash/assets/615622/ffdb3aa2-37b2-4bd2-88ec-a46bebe4a760

- Click on URL field. Click outside. keyboard still visible (unable to unfocus)
- Appbar: Chat symbol for response is not intuitive. Appbar title should be the name of the request (with edit/delete options) rather than just a generic name Requests.
- Rethink the buttons at the bottom of the request page from UX perspective. The send button is smaller than View Code button

227 appbar

- In the collection pane (drawer) clicking on a request does not transition to the request page until the user explicitly swipes left.

227 collection pane

**Comment by DenserMeerkat** [comment]
@animator Thanks for taking the time to review my implementation! I appreciate you pointing out the scrolling and focus problems in the widgets. I too noticed these and I'm already working on solutions to ensure these widgets work flawlessly with touch interactions and deliver the intended functionality.

---

### Response Icon
I spent some time trying to find an icon which represents API Response but couldn't find a particular one so settled with the Comment icon. Will try finding alternatives.
### View Code button
Initially just had the "< >" icon in a squircle shaped button for a clean layout and clear button hierarchy, but added "View Code" labael for user's clarity of the button's function. Happy to revert based on your feedback.

---

**Will get back to you if I have any updates/doubts on fixes.**

**Comment by DenserMeerkat** [comment]
>  (with edit/delete options)

@animator 
If I understood it right, users should  be able to perform all below actions from Appbar?
more_options

**Comment by ashitaprasad** [comment]
Yes, all three should be there.

File: lib/providers/ui_providers.dart [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import 'package:flutter/widgets.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:inner_drawer/inner_drawer.dart';
 
+final mobileDrawerKeyProvider = Provider<GlobalKey<InnerDrawerState>>(
```
Comment (ashitaprasad) [inline lib/providers/ui_providers.dart:5]: Use StateProvider

File: lib/screens/home_page/collection_pane.dart [hunk 2]
```diff
@@ -72,7 +72,6 @@ class CollectionPane extends ConsumerWidget {
           ),
           kVSpacer10,
           Container(
-            height: 30,
             margin: const EdgeInsets.only(right: 8),
```
Comment (ashitaprasad) [inline lib/screens/home_page/collection_pane.dart:75]: Why this padding was removed.

File: lib/screens/home_page/collection_pane.dart [hunk 3]
```diff
@@ -143,14 +142,21 @@ class _RequestListState extends ConsumerState<RequestList> {
     final alwaysShowCollectionPaneScrollbar = ref.watch(settingsProvider
         .select((value) => value.alwaysShowCollectionPaneScrollbar));
     final filterQuery = ref.watch(searchQueryProvider).trim();
+    final isMobile =
+        kIsMobile && MediaQuery.sizeOf(context).width < kMinWindowSize.width;
```
Comment (ashitaprasad) [inline lib/screens/home_page/collection_pane.dart:146]: Is it possible to create an extension over context like context.isMobile
```
context.isMobile = kIsMobile && MediaQuery.sizeOf(context).width < kMinWindowSize.width
```
as this code will be used in a lot of places

File: lib/screens/home_page/collection_pane.dart [hunk 4]
```diff
@@ -166,6 +172,19 @@ class _RequestListState extends ConsumerState<RequestList> {
               },
               itemBuilder: (context, index) {
                 var id = requestSequence[index];
+                if (kIsMobile) {
+                  return ReorderableDelayedDragStartListener(
```
Comment (ashitaprasad) [inline lib/screens/home_page/collection_pane.dart:176]: Why used `ReorderableDelayedDragStartListener`

File: lib/screens/mobile/left_drawer.dart
Comment (ashitaprasad) [inline lib/screens/mobile/left_drawer.dart:1]: Create a `widgets` folder inside `mobile` for mobile related widgets. All Stateful & Stateless widgets will go there.

File: lib/screens/mobile/requests_page/requests_page.dart
Comment (ashitaprasad) [inline lib/screens/mobile/requests_page/requests_page.dart:1]: What is the purpose of this folder `requests_page`?

File: lib/screens/mobile/requests_page/response_drawer.dart
Comment (ashitaprasad) [inline lib/screens/mobile/requests_page/response_drawer.dart:1]: Why is `ResponseDrawer` inside `requests_page`?

File: lib/screens/mobile/requests_page/requests_page.dart
Comment (ashitaprasad) [reply lib/screens/mobile/requests_page/requests_page.dart:1]: In desktop version there is no such folder.

File: lib/utils/http_utils.dart [hunk 5]
```diff
@@ -7,14 +7,14 @@ import '../models/models.dart';
 import 'convert_utils.dart' show rowsToMap;
 import '../consts.dart';
 
-String getRequestTitleFromUrl(String? url) {
+String getRequestTitleFromUrl(String? url, {bool capitalize = false}) {
   if (url == null || url.trim() == "") {
```
Comment (ashitaprasad) [inline lib/utils/http_utils.dart:11]: Why this unnecessary change?

File: lib/widgets/dropdowns.dart [hunk 6]
```diff
@@ -29,10 +31,11 @@ class DropdownButtonHttpMethod extends StatelessWidget {
         return DropdownMenuItem<HTTPVerb>(
           value: value,
           child: Padding(
-            padding: const EdgeInsets.only(left: 16),
+            padding: EdgeInsets.only(left: isMobile ? 8 : 16),
             child: Text(
               value.name.toUpperCase(),
               style: kCodeStyle.copyWith(
+                fontSize: isMobile ? 13 : null,
```
Comment (ashitaprasad) [inline lib/widgets/dropdowns.dart:38]: Why font size is being hard coded here just for dropdown .. it should properly flow from consts.dart?

File: lib/widgets/intro_message.dart [hunk 7]
```diff
@@ -23,6 +23,8 @@ class IntroMessage extends StatelessWidget {
     return FutureBuilder(
       future: introData(),
       builder: (BuildContext context, AsyncSnapshot<void> snapshot) {
+        final isMobile = kIsMobile &&
+            MediaQuery.sizeOf(context).width < kMinWindowSize.width;
```
Comment (ashitaprasad) [inline lib/widgets/intro_message.dart:27]: again .. needs to be shortened everywhere by defining an extension over context.

File: lib/screens/home_page/collection_pane.dart (hunk 2 above)
Comment (DenserMeerkat) [reply lib/screens/home_page/collection_pane.dart:75]: image

- Having a fixed height caused this offset due to TextField's internal contentPadding in the mobile device I tested.
- The height of the field was very narrow and difficult to focus on tap. 
Hence for a fair tap target size I removed the height.
( Also I couldn't find any difference on removing the height in desktop devices too, so felt it was redundant and let the textfield itself take suitable height )

File: lib/screens/home_page/collection_pane.dart (hunk 4 above)
Comment (DenserMeerkat) [reply lib/screens/home_page/collection_pane.dart:176]: Having `ReorderableDragStartListener` in touch devices causes drag event to reorder request which collides with the scroll functionality required if there are a large number of requests in the request list. Hence used the `ReorderableDelayedDragStartListener` to avoid this collision of gestures.

File: lib/utils/http_utils.dart (hunk 5 above)
Comment (DenserMeerkat) [reply lib/utils/http_utils.dart:11]: Felt the lowercase 'untitled' Appbar title didn't look nice, but as you said yes it was an unnecessary change will remove it.

File: lib/app.dart [hunk 8]
```diff
@@ -105,6 +105,8 @@ class DashApp extends ConsumerWidget {
   Widget build(BuildContext context, WidgetRef ref) {
     final isDarkMode =
         ref.watch(settingsProvider.select((value) => value.isDark));
+    final isLargeMobile =
+        MediaQuery.sizeOf(context).width > kMinWindowSize.width;
```
Comment (ashitaprasad) [inline lib/app.dart:109]: can be moved to new `context_extensions.dart`

File: lib/screens/home_page/collection_pane.dart [hunk 9]
```diff
@@ -4,6 +4,7 @@ import 'package:apidash/providers/providers.dart';
 import 'package:apidash/widgets/widgets.dart';
 import 'package:apidash/models/models.dart';
 import 'package:apidash/consts.dart';
+import 'package:apidash/extensions/extensions.dart' show MediaQueryExtension;
```
Comment (ashitaprasad) [inline lib/screens/home_page/collection_pane.dart:7]: You can directly import 'package:apidash/extensions/extensions.dart' everywhere, no need to show MediaQueryExtension.

File: lib/screens/home_page/collection_pane.dart (hunk 2 above)
Comment (ashitaprasad) [reply lib/screens/home_page/collection_pane.dart:76]: Thanks for the deep dive 👍

File: lib/screens/home_page/collection_pane.dart (hunk 4 above)
Comment (ashitaprasad) [reply lib/screens/home_page/collection_pane.dart:175]: This is a very good technical blog topic ... you should spin it in to a short blog.

File: lib/screens/home_page/editor_pane/editor_request.dart [hunk 10]
```diff
@@ -4,22 +4,44 @@ import 'package:apidash/providers/providers.dart';
 import 'package:apidash/consts.dart';
 import 'details_card/details_card.dart';
 import 'url_card.dart';
+import 'package:apidash/screens/home_page/editor_pane/details_card/request_pane/request_pane.dart';
```
Comment (ashitaprasad) [inline lib/screens/home_page/editor_pane/editor_request.dart:7]: This import needs cleaning up
